pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -557,110 +557,200 @@ class Utils {
+                    isJobChanged = true
                 }
-                j.addProperty(p)
+
+                if(isJobChanged){
+                    bc.commit()
+                }
+            } finally {
+                bc.abort()
             }
+        }
+    }
+
+    /**
+     * Compare lists of {@link ParameterDefinition}.
+     *
+     * @param first  First list of parameter definitions.
+     * @param second Second list of parameter definitions.
+     *
+     * @return {@code true}, if both lists of {@link ParameterDefinition} are contain same elements;
+     * {@code false} otherwise
+     */
+    private static boolean isParametersListEquals(@Nonnull List<ParameterDefinition> first,
+                                                  @Nonnull List<ParameterDefinition> second) {
+        if(first.size() != second.size()){
+            return false
+        }
+        Map<String, ParameterDefinition> firstNameToParameterMap  = first.collectEntries {[(it.name): it]}
+        Map<String, ParameterDefinition> secondNameToParameterMap = second.collectEntries {[(it.name): it]}
+        // Check that all parameter's names are equals.
+        if(firstNameToParameterMap.keySet() != secondNameToParameterMap.keySet()){
+            return false
+        }
+        // Check until first non-equal parameter.
+        return !firstNameToParameterMap
+                .values()
+                .any {firstParam -> !isParametersEquals(firstParam, secondNameToParameterMap[firstParam.name])}
+    }
 
-            bc.commit()
-            // Add the action tracking what we added (or empty otherwise)
-            j.replaceAction(new DeclarativeJobPropertyTrackerAction(rawJobProperties, rawTriggers, rawParameters, rawOptions))
-        } finally {
-            bc.abort()
+    /**
+     * Compare {@link ParameterDefinition} objects.
+     *
+     * @param first  First parameter definition.
+     * @param second Second parameter definition.
+     *
+     * @return {@code true}, if both {@link ParameterDefinition} objects are equals; {@code false} otherwise.
+     */
+    private static boolean isParametersEquals(@Nonnull ParameterDefinition first, @Nonnull ParameterDefinition second) {
+        if(first.descriptor.id != second.descriptor.id || first.description != second.description){
+            return false
         }
+        return isObjectsEqualsXStream(first, second)
+    }
+
+    /**
+     * Compare objects' string representations after conversion to XML
+     *
+     * Used for objects:
+     * - all fields are unknown before comparison
+     * - without overwritten methods equals() and hashCode()
+     *
+     * @param first  First object for comparing
+     * @param second Second object for comparing
+     *
+     * @return {@code true} if string representation of objects in XML format are equals; {@code false} otherwise
+     */
+    private static boolean isObjectsEqualsXStream(@Nonnull Object first, @Nonnull Object second) {
```

Comment:
I understand that this works even if the Trigger classes do not implement `getHashCode` or `equals`, but this seems like a really heavy way to check for equality.  Those classes should provide those implementations, right?  

This works but wouldn't you agree it would be better to fix those classes?

Reply:
> Those classes should provide those implementations, right?

Yes, they should!

> This works but wouldn't you agree it would be better to fix those classes?

Yes, they would. But those methods should be implemented in child classes of Trigger, ParameterDefinition, and JobProperty across all plugins. How can we do this?

The suggested method seems heavy at first look, but now all job properties are serialized and written to disk at each run. Approximately this solution will use the same memory resources, but will not block each thread for slow IO.

Reply:
@gulyaev13 
> Yes, they would. But those methods should be implemented in child classes of Trigger, ParameterDefinition, and JobProperty across all plugins. How can we do this?

It will take some time and effort, basically contributing fixes to a number of plugins. It is fine if you want to file a separate JIRA for them, but the goal should be to be to not depend on serialized XML strings for equality comparison.

How about this: use `equals()` comparison between items.  If that returns `true` accept the items as unchanged.  If they return false, fall back to checking `isObjectsEqualsXStream()`.  This should result in better performance for class that define `equals()/hashCode()`, but still working for those that don't.   

We'll need a test for each path - with and without.  But then we could also test the performance difference.